### PR TITLE
Directly start foreground service

### LINKED ISSUE #2010: Context.startForegroundService() did not then call Service.startForeground()
Highest number of reports for ANRs since 2.0.0 in the Dev console - Android 8 only - >5K reports in 60 days

Reason I found google-ing: https://stackoverflow.com/questions/44425584/context-startforegroundservice-did-not-then-call-service-startforeground

> From Google's docs on Android 8.0 behavior changes:
> The system allows apps to call Context.startForegroundService() even while the app is in the background. However, the app must call that service's startForeground() method within five seconds after the service is created.

--> https://developer.android.com/about/versions/oreo/android-8.0-changes.html

************ CAUSE OF ERROR ************

android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{fb59be8 u0 com.nextcloud.client/com.owncloud.android.services.OperationsService}
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1745)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:193)
at android.app.ActivityThread.main(ActivityThread.java:6718)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

************ APP INFORMATION ************
ID: com.nextcloud.client
Version: 30080190
Build flavor: generic

************ DEVICE INFORMATION ************
Brand: lge
Device: joan
Model: LG-H932
Id: PQ3A.190605.003
Product: joan_tmo_us

************ FIRMWARE ************
SDK: 28
Release: 9
Incremental: eng.markus.20190610.202830

### LINKED ISSUE #6905: Crash due to PlayerService
### Steps to reproduce
1. delete a file
2. see 
```
2020-09-08 07:20:21.303 com.nextcloud.client E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.nextcloud.client, PID: 10750
    android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{a08f83f u0 com.nextcloud.client/.media.PlayerService}
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

@ezaquarii this happens on latest dev, and it seems that it is from your foreground fix:
Upon file removed event, FileDisplayActivity onRemoveFileOperationFinish is called and this calls tryStopPlaying.

PR was #6798

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
Fix #2010 
Fix #6905

This is not done at all, but it prevents from crashing on my test device.
@ezaquarii for some reason the startForeground in PlayerService is not called, when starting a video.
Thus it get killed by Android after a specific time.

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### Testing 
Writing tests is very important. Please try to write some tests for your PR. 
If you need help, please do not hesitate to ask in this PR for help.

[unit tests](https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#unit-tests)
[instrumented tests](https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#instrumented-tests)
[UI tests](https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#ui-tests)

- [ ] Tests written, or not not needed

### CODE REVIEW

**Comment by ezaquarii** [comment]
@tobiasKaminsky this is very strange edge case as video player does not use PlayerService at all - it depends on a media player in VideoView.

It sounds to me like the player service might be receiving unintended events from GUI.

**Comment by InfamousUser** [comment]
Tried it, it does not seem to crash, but I think crashing for me is caused by autoupload, which doesn't seem to want to start on this app. I did turn off doze for it.

**Comment by InfamousUser** [comment]
Ok, autoupload started working, it took a while to initialize and well...

The problem is fixed. The problem being the app crash, now the app doesn't crash but instead the whole app or Android system freezes and/or ANR dialogue is shown. This fixed one of the symptoms apparently. Looks like the underlying problem was just exacerbating the crash case, I did not know there was an underlying problem in the first place.

Okay, so the problem for me is app freezing.
Also the ANR dialogue is different now, there's only 2 options whereas there were 3 before.

P.S. Can't even open the app anymore - it stops responding on the initial loading screen.

File: src/main/java/com/nextcloud/client/media/PlayerService.kt [hunk 1]
```diff
@@ -101,6 +101,13 @@ class PlayerService : Service() {
         stop.action = ACTION_STOP
         val pendingStop = PendingIntent.getService(this, 0, stop, 0)
         notificationBuilder.addAction(0, "STOP", pendingStop)
+
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
```
Comment (ezaquarii) [inline src/main/java/com/nextcloud/client/media/PlayerService.kt:105]: This is nasty. But if we need this workaround, I'd put it in `onCreate()`.

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky this is very strange edge case as video player does not use PlayerService at all - it depends on a media player in VideoView.
> 
> It sounds to me like the player service might be receiving unintended events from GUI.

I was confused by that first, too.
When you click on an image it opens PreviewImageFragment
When you click on an audio/video it opens PreviewMediaFragment

PreviewMediaFragment then can do both: play audio or video.
If audio plays and you click "back", then audio continues to play in background (via PlayerService).
I guess that this is then the reason why PlayerService is started.

When you then click on video in PreviewMediaFragment it opens up PreviewVideoActivity to display video in fullscreen…

File: src/main/java/com/nextcloud/client/media/PlayerService.kt (hunk 1 above)
Comment (tobiasKaminsky) [reply src/main/java/com/nextcloud/client/media/PlayerService.kt:105]: True, this is not meant to be the final version.
I just did a quick PR for testing, so that user can give feedback.

**Comment by tobiasKaminsky** [comment]
Reason why PlayerService is started is:
https://github.com/nextcloud/android/blob/77751da6b8cc59dec2bf59941c407adc62aa862c/src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java#L326-L329

Which is somehow needed:
- play audio
- go back -> audio continues in background
- start a video

**Comment by tobiasKaminsky** [comment]
@InfamousUser can you test it again, please?

**Comment by tobiasKaminsky** [comment]
@ezaquarii I found the cause and hopefully a proper fix. Can you review it? :-)

**Comment by InfamousUser** [comment]
> @InfamousUser can you test it again, please?

> Ok, autoupload started working, it took a while to initialize and well...
> 
> The problem is fixed. The problem being the app crash, now the app doesn't crash but instead the whole app or Android system freezes and/or ANR dialogue is shown. This fixed one of the symptoms apparently. Looks like the underlying problem was just exacerbating the crash case, I did not know there was an underlying problem in the first place.
> 
> Okay, so the problem for me is app freezing.
> Also the ANR dialogue is different now, there's only 2 options whereas there were 3 before.
> 
> P.S. Can't even open the app anymore - it stops responding on the initial loading screen.

Probably same as above, I can't test because of the original problem. The app is frozen solid on startup for 80 seconds (loading screen), then displays frozen main interface.

**Comment by tobiasKaminsky** [comment]
> Probably same as above, I can't test because of the original problem. The app is frozen solid on startup for 80 seconds (loading screen), then displays frozen main interface.
> 
> It also locks the Android interface half the times for up to half a minute, that shouldn't be possible, probably a bug in Android.

Can you create another issue for this? 
Please also try to get logcat.

**Comment by InfamousUser** [comment]
Not sure whether you wanted the log here or in the new issue, but here it is here just in case. Also not sure this is what you mean by logcat, never used it before, this is what I got by finding an app called that.

[Nextcloud.txt](https://github.com/nextcloud/android/files/5567844/Nextcloud.txt)

**Comment by InfamousUser** [comment]
New issue: https://github.com/nextcloud/android/issues/7378

File: src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java [hunk 2]
```diff
@@ -324,7 +324,9 @@ public void onStart() {
                 mMultiListContainer.setVisibility(View.GONE);
                 mPreviewContainer.setVisibility(View.VISIBLE);
             } else if (MimeTypeUtil.isVideo(file)) {
-                stopAudio();
+                if (mMediaPlayerServiceConnection.isConnected() && mMediaPlayerServiceConnection.isPlaying()) {
+                    stopAudio();
+                }
```
Comment (ezaquarii) [inline src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java:329]: This looks like a solution for this case.

However, because `stopAudio()` results in `Intent` being sent to a service and thus it starts the service, the call is simply not safe to use. I think we should internally call `Service.stopSelf()` if there is nothing to play as well.

**Review by ezaquarii — APPROVED** [review]
This looks like a solution.

File: src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java (hunk 2 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java:329]: In PreviewMediaFragment onStart() I tried to bind to an existing player service.
Reason is:
- start a long audio file
- start a video
- audio stops, video sound is played, but audio notification is still there (and cannot be resumed)

So my quick idea was to bind to audio service when starting video and stop audio service.
But this does not work.

Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java:329]: Fixed with adding a "play/pause" button.

**Comment by tobiasKaminsky** [comment]
:exclamation: Rebased :exclamation:
